### PR TITLE
Add Substrate EVM Adapter application

### PR DESCRIPTION
**Pull request by dastanbeksamatov**
#### Project Abstract

Substrate EVM adapter aims to present an alternative approach to EVM compatibility for Substrate chains. The main goals are to improve developer experience and introduce an approach that requires the least amount of changes to the runtime and client. It does so by leveraging the best parts of multiple existing compatibility solutions and other awesome ecosystem tools.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (DOT, USDC & USDT) address in the application and bank details via email, if applicable).
- [x] I understand that 30% of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by dastanbeksamatov** [comment]
I have read and hereby sign the Contributor License Agreement.

File: applications/substrate-evm-adapter.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+- **Estimated duration:** 9 weeks
+- **FTE:**  1
+- **Costs:** 18,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | MIT |
+| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
```
Comment (PieWol) [inline applications/substrate-evm-adapter.md:172]: As you are developing a pallet which should be able to interact with other pallets I think you need to provide integration tests in addition to the unit tests.

File: applications/substrate-evm-adapter.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| **0e.** | Article | I will publish an article that explains the complete lifecycle and future plans of the project |
+| 1. | EVM Adapter Pallet | We will create a pallet that will be responsible for bridging the gap between the ETH RPC adapter and the Substrate chain. It will have the ability to dispatch FRAME calls, opt-in possibility to execute EVM bytecode, and handle account mapping. Another main responsibility of the pallet will be handling signature verification. Some parts of this pallet can be inspired from `frontier`'s pallet `pallet-ethereum` but needs refinement and some modification. |
+| 2. | Substrate Node | We will create a Substrate node that has two runtimes: with and without the `pallet-evm`. Both will contain `evm-adapter` pallet, but only the one with `pallet-evm` will be able to execute EVM bytecode. This will demonstrate two main use-cases of this pallet. |
+| 3. | Unit tests | We will provide comprehensive unit tests for the pallet. |
```
Comment (PieWol) [inline applications/substrate-evm-adapter.md:177]: this needs to be covered by the default deliverable 0c so please remove it as an extra milestone deliverable.

File: applications/substrate-evm-adapter.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+| 4. | MVP Demo DApp | We will provide a small EVM DApp that demonstrates the integration of the pallet and RPC adapter. This will be a basic web page that can sign and send custom extrinsic with Metamask, read the state of Substrate pallets. |
+| 5. | `docker-compose` | We will provide a `docker-compose` file that will run the whole project. |
+
+## Future Plans
+
+In the future, we plan to explore if we can completely remove the adapter pallet and have the account mapping and transaction converting logic completely in the RPC adapter. This will allow us to have even more lightweight and generic solution, however could present more challenges. A minimal metamask snap that can share the keyring with the adapter could be one of the solutions, but needs more research.
```
Comment (PieWol) [inline applications/substrate-evm-adapter.md:183]: I think the metamask snaps especially by Chainsafe are exactly what you are looking for.

File: applications/substrate-evm-adapter.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| **0e.** | Article | I will publish an article that explains the complete lifecycle and future plans of the project |
+| 1. | EVM Adapter Pallet | We will create a pallet that will be responsible for bridging the gap between the ETH RPC adapter and the Substrate chain. It will have the ability to dispatch FRAME calls, opt-in possibility to execute EVM bytecode, and handle account mapping. Another main responsibility of the pallet will be handling signature verification. Some parts of this pallet can be inspired from `frontier`'s pallet `pallet-ethereum` but needs refinement and some modification. |
+| 2. | Substrate Node | We will create a Substrate node that has two runtimes: with and without the `pallet-evm`. Both will contain `evm-adapter` pallet, but only the one with `pallet-evm` will be able to execute EVM bytecode. This will demonstrate two main use-cases of this pallet. |
```
Comment (PieWol) [inline applications/substrate-evm-adapter.md:176]: How is this Node different from existing EVM chains like Moonbeam?

File: applications/substrate-evm-adapter.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+### Team LinkedIn Profiles (if available)
+
+- https://linkedin.com/in/dastanbek-samatov
+
+## Development Roadmap :nut_and_bolt:
+
+### Overview
+
+- **Total Estimated Duration:** 3.5 months
```
Comment (PieWol) [inline applications/substrate-evm-adapter.md:141]: Doesn't add up with the estimates of the milestones.

**Review by PieWol — CHANGES_REQUESTED** [review]
To my understanding this application tries to do what the Metamask snaps developed by Chainsafe try to do. I'm personally not fond of supporting another project with the same goal, especially as you are planning to completely discard a core piece of this grant application in the long run (adapter pallet). Other than that I left a few comments that should be addressed. Let's see what my colleagues will say.

File: applications/substrate-evm-adapter.md (hunk 4 above)
Comment (dastanbeksamatov) [reply applications/substrate-evm-adapter.md:176]: moonbeam and many other EVM compatible chains built with Frontier have the ETH RPC baked in in the client. They also add a lot pallets, runtime APIs and other changes to the runtime to ensure full EVM compatibility. Integrated ETH RPC complicates the code, really [slows]((https://github.com/polkadot-evm/frontier/issues/1067#issue-1731136963)) down the node, tps takes a hit and makes it hard to benchmark etc.

This deliverable in comparison, will have the most minimal setup: normal Substrate client and runtime with stripped down, lightweight version of the `pallet-evm` and `pallet-evm-adapter`. `rpc-adapter` is a standalone light client and does not have the problems of integrated rpc

**Comment by dastanbeksamatov** [comment]
thanks for the review! 

In general, I described pros and cons for similar in the `Ecosystem Fit` section and I believe there is a need for this kind of solution, as evidenced by [forum](https://forum.polkadot.network/t/wrappedevm-eth-rpc-compatibility-layer/2775) and StackExchange [posts](https://substrate.stackexchange.com/questions/8983/read-pallet-state-using-metamask).

> what the Metamask snaps developed by Chainsafe try to do.

I took a look at their snap as well, and it comes down to the same thing I mentioned for `polkamask` - the need to install a third party plugin, and the fact that it is closely tied to Metamask itself. It does not work in a script with `web3.js` or `ethers.js`, for example

> I'm personally not fond of supporting another project with the same goal, especially as you are planning to completely discard a core piece of this grant application in the long run (adapter pallet)

I agree with this comment and also understand where the confusion comes from. With enabling near to full evm compatibility without the pallet, I meant having such possibility, pallet would remain there as an option. But it would be possible to "write" to the chain without using the adapter pallet. And `snap` is a very low hanging fruit that can be defaulted to if I don't find any better solution. And that `future plans` section is very raw right now, and I expect it will change during the course of this grant's development.

**Comment by dastanbeksamatov** [comment]
> Welcome back and thanks for the application, @dastanbeksamatov.

thanks @semuelle!

> In the future, we plan to explore if we can make the adapter pallet optional

IMO, one of the main disadvantages of current solutions is the fact that they require a lot of changes to the runtime and client. And by having the adapter pallet, we are not completely fixing that. This is why I think it is important to explore options that could enable frictionless EVM compatibility for any Substrate chain. Obviously without sacrificing security of the chain. 

> Are you planning to use this architecture in a project of your own, or are you just building it in the hopes that it will generate enough community interest to maintain itself?

RPC adapter itself will be a community tool that can run a light client of any Substrate chain and expose ETH RPC from it. As I mentioned above, there is enough community interest for this. And the evm adapter pallet, it will also be a public module that chains can use to provide "almost" full EVM compatibility. I can maintain it for the foreseeable future. If it generates enough interest (especially from the polkadot evm community), I would be interested in integrating it to upstream frontier monorepo as a stripped down version of `pallet-evm`.

**Review by semuelle — APPROVED** [review]
Thanks for the clarification, @dastanbeksamatov. Sounds good to me. I will share your application with the rest of the committee.

**Review by keeganquigley — APPROVED** [review]
Hi @dastanbeksamatov thanks for the explanations and also for the presentation you gave me during our call. Based on your previous work, also happy to approve it.

I know you've worked with us a lot but I don't believe you've gone through a KYC phase with us yet, which we now require for all grantees. Could you take a moment to [submit verification](https://in.sumsub.com/idensic/l/#/znT1S7rcS29rMe1A)? Thanks! Once that is done we can merge the grant.

**Comment by dastanbeksamatov** [comment]
> Hi @dastanbeksamatov thanks for the explanations and also for the presentation you gave me during our call. Based on your previous work, also happy to approve it.
> 
> I know you've worked with us a lot but I don't believe you've gone through a KYC phase with us yet, which we now require for all grantees. Could you take a moment to [submit verification](https://in.sumsub.com/idensic/l/#/znT1S7rcS29rMe1A)? Thanks! Once that is done we can merge the grant.

just completed it! thanks!

**Comment by keeganquigley** [comment]
Hi @dastanbeksamatov milestone 1 is about 6 months overdue now, are you still working on this grant? To be honest I don't think it's super necessary anymore with the upcoming Plaza launch, since we'll soon be able to deploy Solidity contracts targeting PVM. Unless you are planning to submit a delivery soon I think we should go ahead and cancel it.

**Comment by dastansam** [comment]
> Hi @dastanbeksamatov milestone 1 is about 6 months overdue now, are you still working on this grant? To be honest I don't think it's super necessary anymore with the upcoming Plaza launch, since we'll soon be able to deploy Solidity contracts targeting PVM. Unless you are planning to submit a delivery soon I think we should go ahead and cancel it.

hey @keeganquigley, yes I am on the verge of submitting it. I've been working on it for the past month. And at least the first milestone is not really about Solidity contracts, more about ETH RPC compatibility or chain abstraction. So I think there is still a need for it even after the Plaza

**Comment by quigl1kd** [comment]
Fair enough, sounds good. I agree that the ETH RPC adapter and the EVM adapter pallet could still come in handy, at least until the JAM chain is implemented, which will ideally use smoldot, but that's a ways off. As long as you're actively working on it, happy to keep it open :)

**Comment by keeganquigley** [comment]
Fair enough, sounds good. I agree that the ETH RPC adapter and the EVM adapter pallet could still come in handy, at least until the JAM chain is implemented, which will ideally use smoldot, but that's a ways off. As long as you're actively working on it, happy to keep it open :)

**Comment by dastansam** [comment]
> Fair enough, sounds good. I agree that the ETH RPC adapter and the EVM adapter pallet could still come in handy, at least until the JAM chain is implemented, which will ideally use smoldot, but that's a ways off. As long as you're actively working on it, happy to keep it open :)

there is also a need for a not bloated ETH RPC compatibility solution for chains outside of Polkadot relaychain (which I assume Plaza will cover), f.e solochains or any other type of chain build by Substrate.

thanks for understanding, I will make sure to submit the milestones ASAP

**Comment by keeganquigley** [comment]
Thanks @dastansam please let us know if you will have further delays in submitting it.

**Comment by dastansam** [comment]
> Thanks @dastansam please let us know if you will have further delays in submitting it.

I will submit an amendment if I can't do it by the end of the week

**Comment by keeganquigley** [comment]
Hi @dastanbeksamatov how is milestone 2 coming along?

**Comment by dastansam** [comment]
> Hi @dastanbeksamatov how is milestone 2 coming along?

it's WIP, hopefully in couple of weeks

**Comment by keeganquigley** [comment]
Hi @dastanbeksamatov any updates here?

**Comment by dastanbeksamatov** [comment]
> Hi @dastanbeksamatov any updates here?

testing it locally rn. Anytime soon. Thanks!

**Pull request merged**